src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Hover/DefaultRazorHoverInfoService.cs

```diff
@@ -0,0 +1,167 @@
+using HoverModel = OmniSharp.Extensions.LanguageServer.Protocol.Models.Hover;
+using RangeModel = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Hover
+{
+    internal class DefaultRazorHoverInfoService : RazorHoverInfoService
+    {
+        private readonly TagHelperFactsService _tagHelperFactsService;
+        private readonly TagHelperDescriptionFactory _tagHelperDescriptionFactory;
+        private readonly HtmlFactsService _htmlFactsService;
+
+        [ImportingConstructor]
+        public DefaultRazorHoverInfoService(
+            TagHelperFactsService tagHelperFactsService,
+            TagHelperDescriptionFactory tagHelperDescriptionFactory,
+            HtmlFactsService htmlFactsService)
+        {
+            if (tagHelperFactsService is null)
+            {
+                throw new ArgumentNullException(nameof(tagHelperFactsService));
+            }
+
+            if (tagHelperDescriptionFactory is null)
+            {
+                throw new ArgumentNullException(nameof(tagHelperDescriptionFactory));
+            }
+
+            if (htmlFactsService is null)
+            {
+                throw new ArgumentNullException(nameof(htmlFactsService));
+            }
+
+            _tagHelperFactsService = tagHelperFactsService;
+            _tagHelperDescriptionFactory = tagHelperDescriptionFactory;
+            _htmlFactsService = htmlFactsService;
+        }
+
+        public override HoverModel GetHoverInfo(RazorCodeDocument codeDocument, SourceSpan location)
+        {
+            if (codeDocument is null)
+            {
+                throw new ArgumentNullException(nameof(codeDocument));
+            }
+
+            var syntaxTree = codeDocument.GetSyntaxTree();
+            var change = new SourceChange(location, "");
+            var owner = syntaxTree.Root.LocateOwner(change);
+
+            if (owner == null)
+            {
+                Debug.Fail("Owner should never be null.");
+                return null;
+            }
+
+            var parent = owner.Parent;
+            var position = new Position(location.LineIndex, location.CharacterIndex);
+            var tagHelperDocumentContext = codeDocument.GetTagHelperContext();
+
+            if (_htmlFactsService.TryGetElementInfo(parent, out var containingTagNameToken, out var attributes) &&
+                containingTagNameToken.Span.IntersectsWith(location.AbsoluteIndex))
+            {
+                // Hovering over HTML tag name
+                var stringifiedAttributes = _tagHelperFactsService.StringifyAttributes(attributes);
+                var binding = _tagHelperFactsService.GetTagHelperBinding(
+                    tagHelperDocumentContext,
+                    containingTagNameToken.Content,
+                    stringifiedAttributes,
+                    parentTag: null,
```

Comment:
Will this not lead to false positives? What if this is not actually a tag helper because of a parent tag constraint?